sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala

```diff
@@ -22,10 +22,34 @@ import java.time.chrono.IsoChronology
 import java.time.format.{DateTimeFormatter, DateTimeFormatterBuilder, ResolverStyle}
 import java.time.temporal.{ChronoField, TemporalAccessor, TemporalQueries}
 import java.util.Locale
+import java.util.concurrent.{Callable, TimeUnit}
+
+import com.google.common.cache.CacheBuilder
+import com.google.common.util.concurrent.{ExecutionError, UncheckedExecutionException}
 
 trait DateTimeFormatterHelper {
+  protected def toInstantWithZoneId(temporalAccessor: TemporalAccessor, zoneId: ZoneId): Instant = {
+    val localTime = if (temporalAccessor.query(TemporalQueries.localTime) == null) {
+      LocalTime.ofNanoOfDay(0)
+    } else {
+      LocalTime.from(temporalAccessor)
+    }
+    val localDate = LocalDate.from(temporalAccessor)
+    val localDateTime = LocalDateTime.of(localDate, localTime)
+    val zonedDateTime = ZonedDateTime.of(localDateTime, zoneId)
+    Instant.from(zonedDateTime)
+  }
+}
+
+object DateTimeFormatterHelper {
```

Comment:
How about making this private, and make a function under `DateTimeFormatterHelper` trait that called this object? In that way, we won't need 

```scala
class ... extends DateTimeFormatterHelper {
  DateTimeFormatterHelper.getFormatter
}
```

to

```scala
class ... extends DateTimeFormatterHelper {
 getFormatter
}
```

Reply:
I'll move `getOrCreateFormatter` to the trait